decisions/01-new-animation-api.md

```diff
@@ -0,0 +1,208 @@
+- `engine-svg` and `engine-canvas` have no best practices since these are just slim wrappers around the HTML5 elements.
+
+One important aspect of this new animation API is that we want the ability to show a timeline scrubber in the Mechanic UI to give users an easy way to preview a specific frame of their design function output. This is much easier for frameworks that encourage functional design functions such as React and harder for frameworks such as p5.js. This timeline functionality is not described in this decision, but will build upon any decisions made in this proposal.
+
+## Decision
+
+We propose a new animation API where design functions default to a frame-based approach. If a design function is animated, it will automatically get called continously in a pace determined by the `frameRate` setting, until it calls the `done` function to stop the animation. This means that design functions are expected to be pure functions and not rely on global state to render the given frame. This in turn makes it possible for the Mechanic UI to provide a timeline that will update the design function with the given `frameCount` and result in a preview for that specific frame. Note that this new animation API does not impose any rules about the duration of the animation. This is always controlled by calling `done` at some point during the animation lifecycle.
+
+## Implementation Details
+
+The new animation API comes with a few changes to the settings:
+
+- The `animated` setting is renamed to `mode` (default: `static`) and the options are `static` (for static images), `animation` (for the new animation API) and `customAnimation` (for a user to implement their own animation code using the `frame` and `done` callbacks).
+- `frameRate` (default: `60`) is a number that can be used to change the number of frames per second that the design function is called in `animation` mode.
+
+Since the new animation API encourages pure functions, the engines now ship with a `memo` function that can be used to cache things across function calls. Engine-specific helpers such as a `useCanvas` function will also be rewritten to create the canvas element on the first call and return it on following calls.
+
+The argument syntax is also changing slightly by placing the `frame` and `done` functions as root properties of the design function argument object. The `frame` function will only get passed to the design function in `customAnimation` mode.
+
+Here's a look at what this new animation API will look like for each engine.
+
+### `engine-canvas`
+
+```js
+export const handler = async ({ inputs, frameCount, done, useCanvas }) => {
+  const canvas = useCanvas(inputs.width, inputs.height);
+  if (frameCount >= 100) {
+    done(canvas);
```

Comment:
Maybe this is very engine specific, but does done have to receive the canvas? what happens with the one returned right after? I guess that's something that sticks out to me in this new way of writing design functions.

EDIT: now I see you commented on this too in another section. I agree with something like that should also be applied here.